File: backend/services/calculator.py
"""
Financial ratio calculator — all ratios for an Indian SME context.
All monetary inputs assumed to be in Lakhs (INR).
"""
from typing import Optional
from models.financial_data import FinancialData, FinancialRatios


def safe_div(numerator: float, denominator: float, default=None) -> Optional[float]:
    if denominator is None or denominator == 0:
        return default
    return numerator / denominator


def calculate_ratios(data: FinancialData, prev_data: Optional[FinancialData] = None) -> FinancialRatios:
    bs = data.balance_sheet
    pl = data.profit_loss
    cf = data.cash_flow

    revenue = pl.revenue_from_operations
    total_revenue = pl.total_revenue
    gross_profit = pl.gross_profit
    ebitda = pl.ebitda
    ebit = pl.ebit
    pbt = pl.pbt
    pat = pl.pat

    total_assets = bs.total_assets
    total_equity = bs.total_equity
    current_assets = bs.total_current_assets
    current_liabilities = bs.total_current_liabilities
    total_debt = bs.long_term_borrowings + bs.short_term_borrowings

    # Averages for efficiency ratios (use prev year if available)
    if prev_data:
        avg_assets = (total_assets + prev_data.balance_sheet.total_assets) / 2
        avg_equity = (total_equity + prev_data.balance_sheet.total_equity) / 2
        avg_inventory = (bs.inventories + prev_data.balance_sheet.inventories) / 2
        avg_trade_receivables = (bs.trade_receivables + prev_data.balance_sheet.trade_receivables) / 2
        avg_trade_payables = (bs.trade_payables + prev_data.balance_sheet.trade_payables) / 2
    else:
        avg_assets = total_assets
        avg_equity = total_equity
        avg_inventory = bs.inventories
        avg_trade_receivables = bs.trade_receivables
        avg_trade_payables = bs.trade_payables

    capital_employed = total_assets - current_liabilities

    # ── Profitability ──────────────────────────────────────────────
    gross_margin = safe_div(gross_profit, revenue) * 100 if revenue else None
    net_margin = safe_div(pat, revenue) * 100 if revenue else None
    ebitda_margin = safe_div(ebitda, revenue) * 100 if revenue else None
    ebit_margin = safe_div(ebit, revenue) * 100 if revenue else None
    roe = safe_div(pat, avg_equity) * 100 if avg_equity else None
    roa = safe_div(pat, avg_assets) * 100 if avg_assets else None
    roce = safe_div(ebit, capital_employed) * 100 if capital_employed else None
    eps = safe_div(pat, data.balance_sheet.share_capital / 10) if data.balance_sheet.share_capital else None  # Assuming face value ₹10

    # ── Liquidity ──────────────────────────────────────────────────
    current_ratio = safe_div(current_assets, current_liabilities)
    quick_ratio = safe_div(current_assets - bs.inventories, current_liabilities)
    cash_ratio = safe_div(bs.cash_and_equivalents, current_liabilities)
    working_capital = current_assets - current_liabilities

    # ── Leverage ───────────────────────────────────────────────────
    debt_to_equity = safe_div(total_debt, total_equity)
    debt_ratio = safe_div(total_debt, total_assets)
    interest_coverage = safe_div(ebit, pl.finance_costs)
    # DSCR: (PAT + Depreciation) / (Annual Loan Repayment + Interest)
    annual_repayment = data.annual_loan_repayment if data.annual_loan_repayment else total_debt * 0.15  # estimate 15% p.a.
    dscr_denominator = annual_repayment + pl.finance_costs
    dscr = safe_div(pat + pl.depreciation, dscr_denominator)
    net_debt = total_debt - bs.cash_and_equivalents
    net_debt_to_ebitda = safe_div(net_debt, ebitda)

    # ── Efficiency ─────────────────────────────────────────────────
    cogs = pl.cogs if pl.cogs > 0 else revenue * 0.6  # fallback estimate

    dso = safe_div(avg_trade_receivables, revenue) * 365 if revenue else None
    dpo = safe_div(avg_trade_payables, cogs) * 365 if cogs else None
    inventory_turnover = safe_div(cogs, avg_inventory)
    dio = safe_div(365, inventory_turnover) if inventory_turnover else None
    ccc = (dso or 0) + (dio or 0) - (dpo or 0) if all([dso, dio, dpo]) else None
    asset_turnover = safe_div(revenue, avg_assets)
    fixed_asset_turnover = safe_div(revenue, bs.fixed_assets) if bs.fixed_assets else None
    capital_productivity = safe_div(revenue, capital_employed) if capital_employed else None

    # ── Cash Flow ──────────────────────────────────────────────────
    ocf_margin = safe_div(cf.operating_cf, revenue) * 100 if revenue else None
    fcf = cf.free_cash_flow
    cf_to_debt = safe_div(cf.operating_cf, total_debt) if total_debt else None
    cash_conversion_ratio = safe_div(cf.operating_cf, ebitda) if ebitda else None
    capex_intensity = safe_div(cf.capex, revenue) * 100 if revenue else None

    return FinancialRatios(
        gross_margin=round(gross_margin, 2) if gross_margin is not None else None,
        net_margin=round(net_margin, 2) if net_margin is not None else None,
        ebitda_margin=round(ebitda_margin, 2) if ebitda_margin is not None else None,
        ebit_margin=round(ebit_margin, 2) if ebit_margin is not None else None,
        roe=round(roe, 2) if roe is not None else None,
        roa=round(roa, 2) if roa is not None else None,
        roce=round(roce, 2) if roce is not None else None,
        eps=round(eps, 2) if eps is not None else None,

        current_ratio=round(current_ratio, 2) if current_ratio is not None else None,
        quick_ratio=round(quick_ratio, 2) if quick_ratio is not None else None,
        cash_ratio=round(cash_ratio, 2) if cash_ratio is not None else None,
        working_capital=round(working_capital, 2) if working_capital is not None else None,

        debt_to_equity=round(debt_to_equity, 2) if debt_to_equity is not None else None,
        debt_ratio=round(debt_ratio, 2) if debt_ratio is not None else None,
        interest_coverage=round(interest_coverage, 2) if interest_coverage is not None else None,
        dscr=round(dscr, 2) if dscr is not None else None,
        net_debt=round(net_debt, 2) if net_debt is not None else None,
        net_debt_to_ebitda=round(net_debt_to_ebitda, 2) if net_debt_to_ebitda is not None else None,
        total_debt=round(total_debt, 2),

        dso=round(dso, 1) if dso is not None else None,
        dpo=round(dpo, 1) if dpo is not None else None,
        dio=round(dio, 1) if dio is not None else None,
        ccc=round(ccc, 1) if ccc is not None else None,
        asset_turnover=round(asset_turnover, 2) if asset_turnover is not None else None,
        inventory_turnover=round(inventory_turnover, 2) if inventory_turnover is not None else None,
        fixed_asset_turnover=round(fixed_asset_turnover, 2) if fixed_asset_turnover is not None else None,
        capital_productivity=round(capital_productivity, 2) if capital_productivity is not None else None,

        ocf_margin=round(ocf_margin, 2) if ocf_margin is not None else None,
        fcf=round(fcf, 2) if fcf is not None else None,
        cf_to_debt=round(cf_to_debt, 2) if cf_to_debt is not None else None,
        cash_conversion_ratio=round(cash_conversion_ratio, 2) if cash_conversion_ratio is not None else None,
        capex_intensity=round(capex_intensity, 2) if capex_intensity is not None else None,
    )

File: frontend/lib/calculator.ts
/**
 * Financial ratio calculator — all ratios for Indian SME context.
 * All monetary values assumed to be in Lakhs (₹ Lakhs).
 */
import type {
  BalanceSheet,
  ProfitLoss,
  CashFlow,
  FinancialData,
  FinancialRatios,
  DerivedBS,
  DerivedPL,
} from "./types";

function safeDiv(a: number, b: number | undefined | null): number | undefined {
  if (!b || b === 0) return undefined;
  return a / b;
}

export function deriveBS(bs: BalanceSheet): DerivedBS {
  const total_non_current_assets =
    bs.fixed_assets +
    bs.capital_wip +
    bs.long_term_investments +
    bs.deferred_tax_asset +
    bs.long_term_loans_advances +
    bs.other_non_current_assets;

  const total_current_assets =
    bs.inventories +
    bs.trade_receivables +
    bs.cash_and_equivalents +
    bs.short_term_loans_advances +
    bs.gst_itc_receivable +
    bs.tds_advance_tax_receivable +
    bs.other_current_assets;

  const total_assets = total_non_current_assets + total_current_assets;

  const total_equity =
    bs.share_capital + bs.reserves_surplus + bs.money_received_share_warrants;

  const total_non_current_liabilities =
    bs.long_term_borrowings + bs.deferred_tax_liability + bs.long_term_provisions;

  const total_current_liabilities =
    bs.short_term_borrowings +
    bs.trade_payables +
    bs.gst_payable +
    bs.tds_payable +
    bs.pf_esi_payable +
    bs.advance_from_customers +
    bs.other_current_liabilities;

  const total_liabilities_equity =
    total_equity + total_non_current_liabilities + total_current_liabilities;

  return {
    total_non_current_assets,
    total_current_assets,
    total_assets,
    total_equity,
    total_non_current_liabilities,
    total_current_liabilities,
    total_liabilities_equity,
  };
}

export function derivePL(pl: ProfitLoss): DerivedPL {
  const total_revenue = pl.revenue_from_operations + pl.other_income;
  const gross_profit = pl.revenue_from_operations - pl.cogs;
  const ebitda = gross_profit - pl.employee_expenses - pl.other_expenses;
  const ebit = ebitda - pl.depreciation;
  const pbt = ebit - pl.finance_costs + pl.other_income;
  const pat = pbt - pl.tax_expense;

  return { total_revenue, gross_profit, ebitda, ebit, pbt, pat };
}

export function calculateRatios(
  data: FinancialData,
  prevData?: FinancialData | null
): FinancialRatios {
  const bs = data.balance_sheet;
  const pl = data.profit_loss;
  const cf = data.cash_flow;
  const d = deriveBS(bs);
  const p = derivePL(pl);

  const revenue = pl.revenue_from_operations;
  const cogs = pl.cogs || revenue * 0.6;

  // Averages (use previous year if available)
  const prevBS = prevData?.balance_sheet;
  const prevD = prevBS ? deriveBS(prevBS) : null;
  const avg_assets = prevD
    ? (d.total_assets + prevD.total_assets) / 2
    : d.total_assets;
  const avg_equity = prevD
    ? (d.total_equity + prevD.total_equity) / 2
    : d.total_equity;
  const avg_inventory = prevBS
    ? (bs.inventories + prevBS.inventories) / 2
    : bs.inventories;
  const avg_receivables = prevBS
    ? (bs.trade_receivables + prevBS.trade_receivables) / 2
    : bs.trade_receivables;
  const avg_payables = prevBS
    ? (bs.trade_payables + prevBS.trade_payables) / 2
    : bs.trade_payables;

  const total_debt = bs.long_term_borrowings + bs.short_term_borrowings;
  const capital_employed = d.total_assets - d.total_current_liabilities;

  // Profitability
  const gross_margin = (r(safeDiv(p.gross_profit, revenue) ?? 0, 2) ?? 0) * 100 || undefined;
  const net_margin = (r(safeDiv(p.pat, revenue) ?? 0, 2) ?? 0) * 100 || undefined;
  const ebitda_margin = (r(safeDiv(p.ebitda, revenue) ?? 0, 2) ?? 0) * 100 || undefined;
  const ebit_margin = (r(safeDiv(p.ebit, revenue) ?? 0, 2) ?? 0) * 100 || undefined;
  const roe = avg_equity ? r(p.pat / avg_equity * 100, 2) : undefined;
  const roa = avg_assets ? r(p.pat / avg_assets * 100, 2) : undefined;
  const roce = capital_employed ? r(p.ebit / capital_employed * 100, 2) : undefined;
  const eps = bs.share_capital
    ? r((p.pat * 100000) / (bs.share_capital * 100000 / 10), 2)
    : undefined; // face value ₹10

  // Liquidity
  const current_ratio = r(safeDiv(d.total_current_assets, d.total_current_liabilities), 2);
  const quick_ratio = r(
    safeDiv(d.total_current_assets - bs.inventories, d.total_current_liabilities),
    2
  );
  const cash_ratio = r(
    safeDiv(bs.cash_and_equivalents, d.total_current_liabilities),
    2
  );
  const working_capital = r(d.total_current_assets - d.total_current_liabilities, 2);

  // Leverage
  const debt_to_equity = r(safeDiv(total_debt, d.total_equity), 2);
  const debt_ratio = r(safeDiv(total_debt, d.total_assets), 2);
  const interest_coverage = r(safeDiv(p.ebit, pl.finance_costs), 2);
  const annual_repayment =
    data.annual_loan_repayment > 0
      ? data.annual_loan_repayment
      : total_debt * 0.15;
  const dscr_denom = annual_repayment + pl.finance_costs;
  const dscr = dscr_denom > 0
    ? r((p.pat + pl.depreciation) / dscr_denom, 2)
    : undefined;
  const net_debt = r(total_debt - bs.cash_and_equivalents, 2);
  const net_debt_to_ebitda = r(safeDiv(net_debt ?? 0, p.ebitda), 2);

  // Efficiency
  const dso = revenue ? r((avg_receivables / revenue) * 365, 1) : undefined;
  const dpo = cogs ? r((avg_payables / cogs) * 365, 1) : undefined;
  const inventory_turnover = avg_inventory ? r(cogs / avg_inventory, 2) : undefined;
  const dio = inventory_turnover ? r(365 / inventory_turnover, 1) : undefined;
  const ccc =
    dso !== undefined && dio !== undefined && dpo !== undefined
      ? r(dso + dio - dpo, 1)
      : undefined;
  const asset_turnover = r(safeDiv(revenue, avg_assets), 2);
  const fixed_asset_turnover = bs.fixed_assets ? r(revenue / bs.fixed_assets, 2) : undefined;
  const capital_productivity = capital_employed ? r(revenue / capital_employed, 2) : undefined;

  // Cash Flow
  const ocf_margin = revenue ? r((cf.operating_cf / revenue) * 100, 2) : undefined;
  const fcf = r(cf.operating_cf - cf.capex, 2);
  const cf_to_debt = total_debt ? r(cf.operating_cf / total_debt, 2) : undefined;
  const cash_conversion_ratio = p.ebitda
    ? r(cf.operating_cf / p.ebitda, 2)
    : undefined;
  const capex_intensity = revenue ? r((cf.capex / revenue) * 100, 2) : undefined;

  return {
    gross_margin: pct(gross_margin),
    net_margin: pct(net_margin),
    ebitda_margin: pct(ebitda_margin),
    ebit_margin: pct(ebit_margin),
    roe: pct(roe),
    roa: pct(roa),
    roce: pct(roce),
    eps,
    current_ratio,
    quick_ratio,
    cash_ratio,
    working_capital,
    debt_to_equity,
    debt_ratio,
    interest_coverage,
    dscr,
    net_debt,
    net_debt_to_ebitda,
    total_debt: r(total_debt, 2),
    dso,
    dpo,
    dio,
    ccc,
    asset_turnover,
    inventory_turnover,
    fixed_asset_turnover,
    capital_productivity,
    ocf: r(cf.operating_cf, 2),
    ocf_margin,
    fcf,
    cf_to_debt,
    cash_conversion_ratio,
    capex_intensity,
  };
}

// Helpers
function r(v: number | undefined | null, dp: number): number | undefined {
  if (v === undefined || v === null || isNaN(v)) return undefined;
  return parseFloat(v.toFixed(dp));
}

function pct(v: number | undefined | null): number | undefined {
  if (v === undefined || v === null || isNaN(v)) return undefined;
  return parseFloat(v.toFixed(2));
}
